src/bpfprogram.cpp

```diff
@@ -0,0 +1,73 @@
+#include "bpfprogram.h"
+
+#include <optional>
+#include <stdexcept>
+#include <tuple>
+
+namespace bpftrace {
+
+std::optional<BpfProgram> BpfProgram::CreateFromBytecode(
+    const BpfBytecode &bytecode,
+    const std::string &name,
+    MapManager &maps)
+{
+  if (bytecode.find(name) != bytecode.end())
+  {
+    return BpfProgram(bytecode, name, maps);
+  }
+  return std::nullopt;
+}
+
+BpfProgram::BpfProgram(const BpfBytecode &bytecode,
+                       const std::string &name,
+                       MapManager &maps)
+    : bytecode_(bytecode), maps_(maps), name_(name), code_()
+{
+}
+
+const std::vector<uint8_t> &BpfProgram::getCode()
+{
+  return code_;
+}
+
+void BpfProgram::assemble()
+{
+  if (!code_.empty())
+    return;
+
+  auto &section = bytecode_.at(name_);
+  code_.reserve(section.size());
+  code_.assign(section.data(), section.data() + section.size());
+
+  relocateMaps();
+}
+
+void BpfProgram::relocateMaps()
+{
+  struct bpf_insn *insns = reinterpret_cast<struct bpf_insn *>(code_.data());
+  for (uintptr_t i = 0; i < code_.size() / sizeof(struct bpf_insn); ++i)
```

Comment:
nit: you could do the division outside the loop to prevent re-calculation (the compiler might optimize this out anyway)